events/playbook.go

```diff
@@ -1,11 +1,157 @@
 package events
 
-import "github.com/flanksource/incident-commander/api"
+import (
+	"fmt"
+	"strconv"
+	"strings"
+
+	"github.com/flanksource/commons/collections"
+	"github.com/flanksource/commons/logger"
+	"github.com/flanksource/duty/models"
+	"github.com/flanksource/gomplate/v3"
+	"github.com/flanksource/incident-commander/api"
+	v1 "github.com/flanksource/incident-commander/api/v1"
+	"github.com/flanksource/incident-commander/db"
+	"github.com/flanksource/incident-commander/playbook"
+)
+
+type FilterEnv struct {
+	Component *models.Component `json:"component,omitempty"`
+	Check     *models.Check     `json:"check,omitempty"`
+}
 
 func SavePlaybookRun(ctx *api.Context, event api.Event) error {
-	// TODO:
-	// See if any playbook is listening on this event.
-	// Match the filters
-	// If everything goes ok, save the playbook run.
+	specEvent, ok := eventToSpecEvent[event.Name]
+	if !ok {
+		return nil
+	}
+
+	playbooks, err := playbook.FindPlaybooksForEvent(ctx, specEvent.Class, specEvent.Event)
+	if err != nil {
+		return fmt.Errorf("error fetching playbooks: %w", err)
+	}
+
+	var eventResource FilterEnv
+	switch event.Name {
+	case EventCheckFailed, EventCheckPassed:
+		if err := ctx.DB().Model(&models.Check{}).Where("id = ?", event.Properties["id"]).First(&eventResource.Check).Error; err != nil {
+			return api.Errorf(api.ENOTFOUND, "check(id=%s) not found", event.Properties["id"])
+		}
+
+	case EventComponentStatusHealthy, EventComponentStatusUnhealthy, EventComponentStatusInfo, EventComponentStatusWarning, EventComponentStatusError:
+		if err := ctx.DB().Model(&models.Component{}).Where("id = ?", event.Properties["id"]).First(&eventResource.Component).Error; err != nil {
+			return api.Errorf(api.ENOTFOUND, "component(id=%s) not found", event.Properties["id"])
+		}
+	}
+
+	for _, p := range playbooks {
+		playbook, err := v1.PlaybookFromModel(p)
+		if err != nil {
+			logger.Errorf("error converting playbook model to spec: %s", err)
+			logToJobHistory(ctx, p.ID.String(), err.Error())
+			continue
+		}
+
+		run := models.PlaybookRun{
+			PlaybookID: p.ID,
+			Status:     models.PlaybookRunStatusPending,
+			// Parameters: types.JSONStringMap(req.Params), // TODO: How to decide on the requirement parameters.
+			CreatedBy: &ctx.User().ID,
+		}
+
+		if playbook.Spec.Approval == nil || playbook.Spec.Approval.Approvers.Empty() {
+			run.Status = models.PlaybookRunStatusScheduled
+		}
+
+		switch specEvent.Class {
+		case "canary":
+			// What should be the component or config id?
+			if ok, err := matchResource(eventResource.Check.Labels, map[string]any{"check": eventResource.Check}, playbook.Spec.On.Canary); err != nil {
```

Comment:
why not just take an eventResource object e.g. matchResource(eventResource, playbook.Spec.On.Canary)

Reply:
`matchResource` does not know whether it should use the labels from the check or the component.